source/simulator/core.cc

```diff
@@ -1030,8 +1030,25 @@ namespace aspect
         }
       coupling[x.temperature][x.temperature] = DoFTools::always;
       for (unsigned int c=0; c<parameters.n_compositional_fields; ++c)
-        coupling[x.compositional_fields[c]][x.compositional_fields[c]]
-          = DoFTools::always;
+        {
+          const AdvectionField adv_field (AdvectionField::composition(c));
+          const typename Parameters<dim>::AdvectionFieldMethod::Kind method = adv_field.advection_method(introspection);
+          switch (method)
+            {
+              // Coupling cases
+              case Parameters<dim>::AdvectionFieldMethod::fem_field:
+                coupling[x.compositional_fields[c]][x.compositional_fields[c]]
+                  = DoFTools::always;
+                break;
+              // Non-coupling cases
+              case Parameters<dim>::AdvectionFieldMethod::particles:
+                coupling[x.compositional_fields[c]][x.compositional_fields[c]]
+                  = DoFTools::none;
+                break;
+              default:
+                AssertThrow(false,ExcNotImplemented());
```

Comment:
I was matching the pattern used for updating the field in the core loop. I can see the argument for an Assert being better in this case though.